src/Amazon.AspNetCore.Identity.Cognito/Extensions/CognitoServiceCollectionExtensions.cs

```diff
@@ -96,8 +98,30 @@ public static CognitoUserPool CreateUserPoolClient(IServiceProvider provider)
             }
 
             var cognitoClient = provider.GetService<IAmazonCognitoIdentityProvider>();
+
+            if (cognitoClient is AmazonCognitoIdentityProviderClient eventProvider)
+            {
+                eventProvider.BeforeRequestEvent += ServiceClientBeforeRequestEvent;
+            }
+
             var cognitoPool = new CognitoUserPool(options.UserPoolId, options.UserPoolClientId, cognitoClient, options.UserPoolClientSecret);
             return cognitoPool;
         }
+
+        private static void ServiceClientBeforeRequestEvent(object sender, Amazon.Runtime.RequestEventArgs e)
+        {
+            Amazon.Runtime.WebServiceRequestEventArgs args = e as Amazon.Runtime.WebServiceRequestEventArgs;
+            if (args == null || !args.Headers.ContainsKey(UserAgentHeader))
+                return;
+
+            args.Headers[UserAgentHeader] = args.Headers[UserAgentHeader] + " CognitoASPNETCoreIdentityProvider/" + GetAssemblyFileVersion();
+        }
+
+        private static string GetAssemblyFileVersion()
+        {
+            var assembly = typeof(CognitoServiceCollectionExtensions).GetTypeInfo().Assembly;
+            AssemblyFileVersionAttribute attribute = assembly.GetCustomAttribute(typeof(AssemblyFileVersionAttribute)) as AssemblyFileVersionAttribute;
```

Comment:
Can we cache this value o we are not doing reflection constantly? Can we just make it a readonly variable?

Reply:
Moved it out.